internal/tescale/scale.go

```diff
@@ -0,0 +1,267 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tescale provides functions for tescale
+package tescale
+
+import (
+	"sync"
+
+	"github.com/openconfig/featureprofiles/internal/deviations"
+	"github.com/openconfig/featureprofiles/internal/iputil"
+	"github.com/openconfig/gribigo/fluent"
+	"github.com/openconfig/ondatra"
+)
+
+const (
+	// VRFT vrf t
+	VRFT = "vrf_t"
+	// VRFR vrf r
+	VRFR = "vrf_r"
+	// VRFRD vrf rd
+	VRFRD = "vrf_rd"
+
+	// V4TunnelIPBlock tunnel IP block
+	V4TunnelIPBlock = "198.18.0.1/16"
+	// V4VIPIPBlock vip IP block
+	V4VIPIPBlock = "198.18.196.1/22"
+
+	tunnelSrcIP = "198.18.204.1"
+)
+
+// IPPool for IPs
+type IPPool struct {
+	ips   []string
+	index int
+	rw    sync.RWMutex
+}
+
+// NewIPPool creates a new IPPool
+func NewIPPool(entries []string) *IPPool {
+	return &IPPool{
+		ips:   entries,
+		index: -1,
+	}
+}
+
+// NextIP returns the next IP
+func (p *IPPool) NextIP() string {
+	p.rw.Lock()
+	defer p.rw.Unlock()
+
+	p.index++
+	return p.ips[p.index]
+}
+
+// AllIPs returns all IPs in the pool
+func (p *IPPool) AllIPs() []string {
+	return append([]string{}, p.ips...)
+}
+
+// IDPool for NH and NHG IDs
+type IDPool struct {
+	nhIndex  uint64
+	nhgIndex uint64
+	rw       sync.RWMutex
+}
+
+// NewIDPool creates a new IDPool
+func NewIDPool(base uint64) *IDPool {
+	return &IDPool{
+		nhIndex:  base,
+		nhgIndex: base,
+	}
+}
+
+// NextNHID returns the next NHID
+func (p *IDPool) NextNHID() uint64 {
+	p.rw.Lock()
+	defer p.rw.Unlock()
+
+	p.nhIndex++
+	return p.nhIndex
+}
+
+// NextNHGID returns the next NHGID
+func (p *IDPool) NextNHGID() uint64 {
+	p.rw.Lock()
+	defer p.rw.Unlock()
+
+	p.nhgIndex++
+	return p.nhgIndex
+}
+
+// VRFConfig holds NH, NHG and IPv4 entries for the VRF.
+type VRFConfig struct {
+	Name      string
+	NHs       []fluent.GRIBIEntry
+	NHGs      []fluent.GRIBIEntry
+	V4Entries []fluent.GRIBIEntry
+}
+
+// Param TE holds scale parameters.
+type Param struct {
+	V4TunnelCount         int
+	V4TunnelNHGCount      int
+	V4TunnelNHGSplitCount int
+	EgressNHGSplitCount   int
+	V4ReEncapNHGCount     int
+}
+
+// BuildVRFConfig creates scale new scale VRF configurations.
+func BuildVRFConfig(dut *ondatra.DUTDevice, egressIPs []string, param Param) []*VRFConfig {
+	v4TunnelIPAddrs := NewIPPool(iputil.GenerateIPs(V4TunnelIPBlock, param.V4TunnelCount))
+	v4VIPAddrs := NewIPPool(iputil.GenerateIPs(V4VIPIPBlock, (param.V4TunnelNHGCount*param.V4TunnelNHGSplitCount)+2))
```

Comment:
Looks like we are creating over 500 VIP routes. in prod, the expectation is that there will only be 2 VIP routes in default VRF.

Reply:
@xw-g could you please confirm check and confirm